src/core/compiler/CircomCompilerDownloader.ts

```diff
@@ -0,0 +1,245 @@
+      case "linux":
+        return arch === "arm64" ? CompilerPlatformBinary.LINUX_ARM : CompilerPlatformBinary.LINUX_AMD;
+      case "darwin":
+        return arch === "arm64" ? CompilerPlatformBinary.MACOS_ARM : CompilerPlatformBinary.MACOS_AMD;
+      default:
+        return CompilerPlatformBinary.WASM;
+    }
+  }
+
+  private static _downloaderPerPlatform: Map<string, CompilerDownloader> = new Map();
+
+  public static getConcurrencySafeDownloader(platform: CompilerPlatformBinary, compilersDir: string) {
+    const key = platform + compilersDir;
+
+    if (!this._downloaderPerPlatform.has(key)) {
+      this._downloaderPerPlatform.set(key, new CompilerDownloader(platform, compilersDir));
+    }
+
+    return this._downloaderPerPlatform.get(key)!;
+  }
+
+  private readonly _mutex = new MultiProcessMutex("compiler-download");
+
+  constructor(
+    private readonly _platform: CompilerPlatformBinary,
+    private readonly _compilersDir: string,
+  ) {}
+
+  public isCompilerDownloaded(version: string): boolean {
+    const downloadPath = this._getCompilerDownloadPath(version);
+    const downloadPathWasm = this._getWasmCompilerDownloadPath(version);
+
+    return fs.pathExistsSync(downloadPath) || fs.pathExistsSync(downloadPathWasm);
+  }
+
+  public async isCompatibleCompilerDownloaded(highestVersion: string): Promise<boolean> {
+    const latestDownloadedVersion = await this.getLatestDownloadedCircomVersion();
+
+    const isVersionCompatible = isVersionHigherOrEqual(latestDownloadedVersion, highestVersion);
+
+    return isVersionCompatible && this.isCompilerDownloaded(latestDownloadedVersion);
+  }
+
+  public async getCompilerBinary(version: string): Promise<CompilerPath> {
+    const compilerBinaryPath = this._getCompilerDownloadPath(version);
+    const wasmCompilerBinaryPath = this._getWasmCompilerDownloadPath(version);
+
+    if (await fs.pathExists(compilerBinaryPath)) {
+      return { binaryPath: compilerBinaryPath, isWasm: false };
+    }
+
+    if (await fs.pathExists(wasmCompilerBinaryPath)) {
+      return { binaryPath: wasmCompilerBinaryPath, isWasm: true };
+    }
+
+    throw new HardhatZKitError(`Trying to get a Circom compiler v${version} before it was downloaded`);
+  }
+
+  public async getLatestCompilerBinary(): Promise<CompilerPath> {
+    const latestDownloadedCompilerVersion = await this.getLatestDownloadedCircomVersion();
+    if (!latestDownloadedCompilerVersion || latestDownloadedCompilerVersion === "0.0.0") {
+      throw new HardhatZKitError("No latest compiler found");
+    }
+
+    const compilerBinaryPath = this._getCompilerDownloadPath(latestDownloadedCompilerVersion);
+
+    if (fs.existsSync(compilerBinaryPath)) {
+      return { binaryPath: compilerBinaryPath, isWasm: false };
+    }
+
+    return { binaryPath: this._getWasmCompilerDownloadPath(latestDownloadedCompilerVersion), isWasm: true };
+  }
+
+  public async downloadCompiler(version: string, isVersionStrict: boolean): Promise<void> {
```

Comment:
```suggestion
  public async downloadCompiler(version?: string): Promise<void> {
```

If version is passed, you treat it as strict. Otherwise, use latest